### ISSUE TITLE
Support for fixed content areas like a footer

### ISSUE THREAD

**Issue by joho1968** [madprops/Msg#9]
Nice job!

It'd be nice if there was support out-of-the-box for fixed content areas like a footer where one could place buttons, static text, whatever. The footer should never disappear, even if content is scrolled and could have its own set of styles. Just a thought. The total "window" size obviously have to take a footer into account so that the footers don't disappear from the browser viewport.

**Comment by madprops** [comment]
Hello. That's a weird idea, why would a "window manager" provide the ability to create fixed elements like footers? That sounds like something you'd want to do with 'native' html/css layouts, unless there's some good reason I'm not seeing.

**Comment by joho1968** [comment]
Fair point :-) I'm trying to use Msg as a jquery-dialog (modal) replacement, where I'd want something like OK/CANCEL buttons at the bottom. Not necessarily both, and not necessarily always. It could be for an input form, or a file browsing dialog, or other object/item selection dialog.

I think I got this working with some CSS/HTML, but I just think it'd be nice if this could be handled natively by Msg.

**Comment by madprops** [comment]
![2020-10-08_07-27](https://user-images.githubusercontent.com/206498/95465028-e1bc5700-0937-11eb-9fdb-b34c27279a64.png)


```
<script src='msg.js'></script>

<style>
  .footer {
    height: 4rem;
    display: grid;
    grid-auto-flow: column;
    align-items: center;
    grid-gap: 0.2rem;
    background-color: #cecece;
  }

  .footer_item {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    background-color: #a7a7ba;
    height: 100%;
  }

  .Msg-content-withfooter {
    padding: 0 !important;
  }

  .content {
    font-size: 2rem;
    padding: 2rem;
  }
</style>

<script>
  window.onload = function () {
    let win = Msg.factory({
      id: 'test',
      class: 'withfooter',
      enable_titlebar: true,
      show_effect: 'none',
      close_effect: 'none',
      persistent: false
    })

    let html = `
      <div class='content'>Hello World</div>
      <div class='footer'>
        <div class='footer_item'>OK</div>
        <div class='footer_item'>Cancel</div>
      </div>
    `

    win.show(["Test", html])
  }
</script>
```

What you are trying to do can be done by simply removing 'padding' from the .Msg-content element and  then just doing your own layout inside. In this case I gave this the class called 'withfooter' and changed .Msg-content-withfooter to remove the padding (the !important is necessary). Then it's up to you to implement whatever you want inside. I think the content inside the windows should be left to the developers to implement, instead of making Msg more complex. What do you think?

**Comment by joho1968** [comment]
Thanks, yes that will work. Mine looked a bit different, but I do see your point.

````css
#Msg-content-dialog {
    background: white !important;
    padding: 2px !important;
    margin: 0 !important;
    overflow: hidden !important;
    height: calc(100% - 64px) !important;
}
.dialog-msg-body {
    background-color: rgba(255, 0, 0, 0.5) !important;
    text-align: left !important;
    padding: 0px !important;
    margin-bottom: 62px !important;
    min-width: 750px !important;
    min-height: 400px !important;
    overflow-y: scroll !important;
}
.dialog-msg-footer {
    height: 60px !important;
    background-color: rgba(0, 0, 255, 0.5) !important;
    color: white !important;
    position: absolute !important;
    bottom: 0 !important;
    left: 0 !important;
    right: 0 !important;
    padding: 2px !important;
}
````

**Comment by madprops** [comment]
Yeah similar approach. I do understand that having a shortcut for this is useful. I suggest using a templating library like Handlebars or creating a 'factory' function that prepares the html.

**Comment by joho1968** [comment]
On a semi-related note, is there a way to update the window after re-loading content via an ajax call that takes the new content into account when sizing the window, etc?

**Comment by madprops** [comment]
I don't understand your question because changes to the content inside should change the dimensions of the window as it adapts to the content, unless you gave the window or containers explicit width and height. Can you describe what you're doing?

**Comment by joho1968** [comment]
1. Create window with initial content, like "Fetching items, please wait ..."
2. Issue ajax call
3. Based on ajax call, show error message or update content inside window
4. Then I'd like to do a "window.recalc()", or something like it

The content will be different when it comes to height, width, etc. So it'd almost be like a completely new window. But I want to avoid having to create a new window since that'll cause "flashing". The user can then select something in the window which will repeat the process from step one and once again load something via ajax.

**Comment by madprops** [comment]
Is there a reason why you give the window a specific width/height? If you don't specify those, the window will just grow/shrink by itself to adapt to the content, no need for manual recalculation. Maybe if I can see how you're creating the window I'd be able to help more.

**Comment by joho1968** [comment]
The only reason my CSS has min-height and min-width is because I don't want to end up with a "dialog" that is "tiny". Errors are rare, but do occur, in which case it's fine to show the error message in the previous window's dimensions so to speak. It's when content grows between ajax calls that is the biggest concern. I don't have too many factory parameters when creating the Msg window, other than disabling effects.

The content is basically:

````html
<div id="dynamic_content" class="dialog-msg-body">
    [dynamic content]
</div>
<div class="dialog-msg-footer">
    Footer content
</div>
````

After the ajax call, #dynamic_content is updated dynamically (i.e. its contents).

Perhaps there's a better way of doing this with Msg.

**Comment by madprops** [comment]
>The only reason my CSS has min-height and min-width 
I don't have too many factory parameters when creating the Msg window, other than disabling effects.

Sounds like you're doing this to your own content container, and not handled by Msg.

>It's when content grows between ajax calls that is the biggest concern.

If you only give it MIN width/height then it shouldn't be a problem.

**Comment by madprops** [comment]
Ok I'm closing this now.
